### PR TITLE
Added preprocessor mask_multimodel

### LINKED ISSUE #487: New preprocessor mask_multimodel
**Is your feature request related to a problem? Please describe.**
I want to propagate the mask over several datasets such that when an element is masked in one dataset, it is also masked in all other datasets.

**Would you be able to help out?**
Yes, working on it. Initially I thought `mask_fillvalues` would do the job, but it doesn't. Also Github kindly points me to other issues, and just now I realize this has been discussed before here https://github.com/ESMValGroup/ESMValCore/issues/116

### PR DESCRIPTION
**Pull request by schlunma**
This PR adds a working multimodel masking preprocessor.

The following things need to be implemented before the draft status can be cleared:

- [x] Add documentation (and remove the old, wrong parts)
- [x] Add unit tests
- [x] Discuss about the name of the preprocessor, I do not think that `mask_multimodel` is the optimal choice.
- [x] Adapt provenance of multimodel output

I probably do not have time to finish this soon, so any contributions to this PR are welcome!

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] If writing a new/modified preprocessor function, please update the [documentation](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/recipe/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes
-   [x] If you make backward incompatible changes to the recipe format, make a new pull request in the [ESMValTool repository](https://github.com/ESMValGroup/ESMValTool) and add the link below

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #487.

### CODE REVIEW

**Review by bascrezee — APPROVED** [review]
To me these changes look good. I also tested this preprocessor quite thoroughly by now, and it works as expected.

**Comment by valeriupredoi** [comment]
@schlunma you wanna switch this PR from Draft to RfR, mate? I'll haz a look right after you did that :+1:

**Comment by schlunma** [comment]
I still need to add tests and documentation for this, will do that this week/next week. After that I will switch the status of this PR :+1:

**Comment by valeriupredoi** [comment]
> I still need to add tests and documentation for this, will do that this week/next week. After that I will switch the status of this PR +1

rock on, dude! :metal:

**Comment by schlunma** [comment]
This is ready for review now :+1: 

I added unit tests and documentation and also tested it on real data with the recipe mentioned in #487. All works as expected.

File: esmvalcore/preprocessor/_mask.py [hunk 1]
```diff
@@ -527,6 +528,97 @@ def mask_outside_range(cube, minimum, maximum):
...
+        for product in products:
+            cubes.extend(product.cubes)
+        if not cubes:
+            return products
+        shape = _get_shape(cubes)
+        return _multimodel_mask_products(products, shape)
+    raise TypeError(
+        f"Input type for mask_multimodel not understood. Expected "
+        f"iris.cube.Cube or esmvalcore.preprocessor.PreprocessorFile, "
+        f"got {products}")
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_mask.py:619]: to be absolutely user-friendly maybe better like `{set(type(p) for p in products)}`

File: tests/unit/preprocessor/_mask/test_mask_multimodel.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+def assert_array_equal(array_1, array_2):
+    """Assert that (masked) array 1 equals (masked) array 2."""
+    if np.ma.is_masked(array_1) or np.ma.is_masked(array_2):
+        np.testing.assert_array_equal(np.ma.getmaskarray(array_1),
+                                      np.ma.getmaskarray(array_2))
+        mask = np.ma.getmaskarray(array_1)
+        np.testing.assert_array_equal(array_1[~mask], array_2[~mask])
+    else:
+        np.testing.assert_array_equal(array_1, array_2)
```
Comment (valeriupredoi) [inline tests/unit/preprocessor/_mask/test_mask_multimodel.py:38]: just use the `assert_array_equal` func from our test library, man, no need to reinvent the wheel

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
looks good! a couple minor points inline, but also - could you pls add an integration test inside `test_recipe` where you call the preprocessor and do some stuff with it? Cheers, dude :beer:

File: tests/unit/preprocessor/_mask/test_mask_multimodel.py (hunk 2 above)
Comment (schlunma) [reply tests/unit/preprocessor/_mask/test_mask_multimodel.py:38]: 1. This function can't be used for scalars, e.g., `AttributeError: 'int' object has no attribute 'mask'`.
2. It fails for arrays like `a = np.ma.masked_equal([0, 1], 0)` and `b = np.ma.masked_equal([1, 1], 1)` which it shouldn't do in my opinion. Both arrays are essentially `[-- 1]`.

I don't like to change the function in the library, maybe some other tests depend on this specific behavior and I don't wanna mess around with that.

Comment (valeriupredoi) [reply tests/unit/preprocessor/_mask/test_mask_multimodel.py:38]: hmm maybe we should develop that function a bit more then, anyways not in this PR so fine by me :+1:

**Review by valeriupredoi — APPROVED** [review]
excellent stuff, cheers much @schlunma :beer:

**Pull request merged**